RIOT-OS/RIOT
# examples/gnrc_border_router: add example for WiFi ⇆ esp_now border router

### Contribution description

This adds an example configuration to use the `gnrc_border_router` to route
between a regular WiFi network and an `esp_now` network.

All you need is an esp based board (esp8266, esp32), an IPv6 uplink and a router that supports DHCPv6. Any OpenWRT router or Fritz!Box will do, make sure you have the relavant config opion enabled:
![Fritz!Box](https://user-images.githubusercontent.com/1301112/75827407-e0c3b580-5da9-11ea-97d1-a770ee6b0636.png)

### Testing procedure

Configure the access data for your WiFi network in `Makefile.esp.conf`, then run

    make -C examples/gnrc_border_router BOARD=esp32-wroom-32 UPLINK=wifi PORT=/dev/ttyUSB0 flash term 

to turn a e.g. `esp32-wroom-32` into a boarder router.

You can use other esp based boards as nodes in the network by flashing
e.g. the `gnrc_networking` example.
Be sure to add `CFLAGS += -DGNRC_IPV6_NIB_CONF_SLAAC=1` as otherwise
your esp_now node will not receive a link-local IP address.

    CFLAGS="-DGNRC_IPV6_NIB_CONF_SLAAC=1" make -C examples/gnrc_networking BOARD=esp8266-esp-12x PORT=/dev/ttyUSB1 flash term

It may take a couple of minutes before the border router has a global address on both interfaces, but then you should be able to ping global addresses from the esp_now nodes.

#### `esp32-wroom-32`

```
2020-03-03 23:55:59,520 # ifconfig
2020-03-03 23:55:59,526 # Iface  9  HWaddr: 3C:71:BF:9E:13:FC  Link: up 
2020-03-03 23:55:59,530 #           L2-PDU:1500 MTU:1492  HL:255  RTR  
2020-03-03 23:55:59,533 #           RTR_ADV  Source address length: 6
2020-03-03 23:55:59,536 #           Link type: wireless
2020-03-03 23:55:59,542 #           inet6 addr: fe80::3e71:bfff:fe9e:13fc  scope: link  VAL
2020-03-03 23:55:59,549 #           inet6 addr: 2001:16b8:45c7:d600:3e71:bfff:fe9e:13fc  scope: global  VAL
2020-03-03 23:55:59,552 #           inet6 group: ff02::2
2020-03-03 23:55:59,554 #           inet6 group: ff02::1
2020-03-03 23:55:59,558 #           inet6 group: ff02::1:ff9e:13fc
2020-03-03 23:55:59,559 #           
2020-03-03 23:55:59,562 # Iface  8  HWaddr: 3C:71:BF:9E:13:FD 
2020-03-03 23:55:59,566 #           L2-PDU:249 MTU:1280  HL:64  RTR  
2020-03-03 23:55:59,570 #           RTR_ADV  6LO  Source address length: 6
2020-03-03 23:55:59,573 #           Link type: wireless
2020-03-03 23:55:59,579 #           inet6 addr: fe80::3e71:bfff:fe9e:13fd  scope: link  VAL
2020-03-03 23:55:59,586 #           inet6 addr: 2001:16b8:45c7:d6f4:3e71:bfff:fe9e:13fd  scope: global  VAL
2020-03-03 23:55:59,588 #           inet6 group: ff02::2
2020-03-03 23:55:59,591 #           inet6 group: ff02::1
2020-03-03 23:55:59,595 #           inet6 group: ff02::1:ff9e:13fd
```

#### `esp8266-esp-12x`

```
2020-03-03 23:56:13,765 # Iface  11  HWaddr: 82:7D:3A:7F:E0:1D 
2020-03-03 23:56:13,768 #           L2-PDU:249 MTU:1280  HL:64  RTR  
2020-03-03 23:56:13,773 #           6LO  Source address length: 6
2020-03-03 23:56:13,776 #           Link type: wireless
2020-03-03 23:56:13,782 #           inet6 addr: fe80::807d:3aff:fe7f:e01d  scope: link  VAL
2020-03-03 23:56:13,788 #           inet6 addr: 2001:16b8:45c7:d6f4:807d:3aff:fe7f:e01d  scope: global  VAL
2020-03-03 23:56:13,790 #           inet6 group: ff02::2
2020-03-03 23:56:13,793 #           inet6 group: ff02::1
2020-03-03 23:56:13,796 #           inet6 group: ff02::1:ff7f:e01d
2020-03-03 23:56:13,799 #           inet6 group: ff02::1a
2020-03-03 23:56:13,801 #           
2020-03-03 23:56:13,804 #           Statistics for Layer 2
2020-03-03 23:56:13,807 #             RX packets 0  bytes 0
2020-03-03 23:56:13,810 #             TX packets 0 (Multicast: 0)  bytes 66
2020-03-03 23:56:13,815 #             TX succeeded 1 errors 0
2020-03-03 23:56:13,816 #           Statistics for IPv6
2020-03-03 23:56:13,820 #             RX packets 1  bytes 120
2020-03-03 23:56:13,824 #             TX packets 5 (Multicast: 5)  bytes 288
2020-03-03 23:56:13,828 #             TX succeeded 5 errors 0
2020-03-03 23:56:13,829 # 
> ping6 2600::
2020-03-03 23:56:17,666 #  ping6 2600::
2020-03-03 23:56:17,721 # scandone
2020-03-03 23:56:18,886 # 12 bytes from 2600::: icmp_seq=1 ttl=49 time=213.690 ms
2020-03-03 23:56:19,913 # 12 bytes from 2600::: icmp_seq=2 ttl=49 time=240.123 ms
2020-03-03 23:56:20,667 # 
2020-03-03 23:56:20,668 # --- 2600:: PING statistics ---
2020-03-03 23:56:20,673 # 3 packets transmitted, 2 packets received, 33% packet loss
2020-03-03 23:56:20,679 # round-trip min/avg/max = 213.690/226.906/240.123 ms
```

### Issues/PRs references

The border router is not very reliable. After a short while, routing may stop working, but it may also start working agaim.

```
2020-03-03 23:58:27,417 # ping6 2600::
2020-03-03 23:58:27,744 # scandone
2020-03-03 23:58:30,418 # 
2020-03-03 23:58:30,419 # --- 2600:: PING statistics ---
2020-03-03 23:58:30,426 # 3 packets transmitted, 0 packets received, 100% packet loss
> 2020-03-03 23:58:37,746 #  scandone
2020-03-03 23:58:59,657 # ping6 2600::
2020-03-03 23:58:59,862 # 12 bytes from 2600::: icmp_seq=0 ttl=49 time=199.205 ms
2020-03-03 23:59:00,884 # 12 bytes from 2600::: icmp_seq=1 ttl=49 time=220.856 ms
2020-03-03 23:59:01,907 # 12 bytes from 2600::: icmp_seq=2 ttl=49 time=244.704 ms
2020-03-03 23:59:01,907 # 
2020-03-03 23:59:01,910 # --- 2600:: PING statistics ---
2020-03-03 23:59:01,915 # 3 packets transmitted, 3 packets received, 0% packet loss
2020-03-03 23:59:01,920 # round-trip min/avg/max = 199.205/221.588/244.704 m
```

There is also the occasional

```
2020-03-04 00:01:25,265 # gnrc_netif: possibly lost interrupt.
2020-03-04 00:01:25,268 # gnrc_netif: possibly lost interrupt.
2020-03-04 00:01:25,272 # gnrc_netif: possibly lost interrupt.
2020-03-04 00:01:25,275 # gnrc_netif: possibly lost interrupt.
``` 

on the border router or 

```
2020-03-04 00:05:35,861 # LmacRxBlk:0
2020-03-04 00:05:36,862 # LmacRxBlk:0
2020-03-04 00:05:37,863 # LmacRxBlk:0
```

on the node.
